text/0000-network-spec.md

```diff
@@ -20,40 +20,96 @@ This document should serve as reference for all the clients to implement network
+- Nodes can be added to the network.
+- Edges can be added and removed to the network.
+- For every known peer it keeps which are the active connections that belong to the shortest route to them.
+- This is recalculated periodically based on last change, and how far from us is the affected node.
+
+Round Robin will be used to pick which peer to send a message in case a routed message is needed.
+
+**Recalculation Policy**
+Recalculation is scheduled when needed.
+
+- If an edge is added between nodes two with absolute distance difference less than 1 no recalculation is needed.
+- If an edge is removed between two nodes and distance to them
+
+Recalculation is scheduled based on the distance of the closest affected peer. The closest the node is, the earlier recalculation is scheduled. We should have `MIN_RECALCULATION_IDLE_TIME` to avoid spending a lot of CPU keeping the shortest path to all nodes synced.
+
+## Peer Manager
+
+### API
+
+- **Broadcast Message** (This will use broadcast idea from )
+- **Send Message to PeerID** (Message will be routed to known PeerID) This will use Find route to peer id from routing table. Messages can travel with a TTL, so they get dropped after lurking around the network. This also avoid loops when the view of the graph is not the same to all nodes.
+- **Send Direct Message PeerID** In this case PeerID must be an active peer.
+- **Ban PeerID**
+
+### Storing information
+
+- Active Peers
+- Banned peers
+- Peer information (This is a long term database where we store information of all peers we have ever met).
+
+## Sync Service
+
+This service will handle keeping the network view for every peer up to date using Broadcast mechanism. This will manage MST. When new edge is created this information is propagated through the network.
+
+Also there is a hard sync mechanism for two nodes, that is used when new connection between two nodes is established.  Periodically nodes will run this hard sync to become up to date. The period to hard sync is always large enough so that changes are allowed to get propagated through the spanning tree.
+If view of the network differ by a large margin nodes will send their full view of the graph. Useful when new nodes join for the first time.
+Otherwise they will use bloom filters to find differences.
+
+**AddEdge/RemoveEdge** messages should contains a timestamp/nonce to avoid situations when removal gets to a node before addition (though in practice this should be extremely rare). When adding a new edge, both nodes after having signature from the other peer start broadcasting this new edge. Use multiple source broadcast mechanism.
+
+## Connecting Service
+
+This service will handle active connections.
+
+- It will determine if an incoming connection should be established. When some node is trying to establish a new connection if peers slots is full we will ping peer with latest last message. If ping fails, we drop our connection and establish connection with new peer.
```

Comment:
hm, that means you are keeping connection open with newly connected peer during ping (what is the timeout on ping?). Eg. attack would be to spam with bunch of connections and which may get node to "too many files open".

Generally TCP/IP I think handles disconnects within reasonable time, so this might not really be needed.

Reply:
Actually I got this idea from Ethereum. Maybe for new incoming connections we should use UDP instead, to avoid wasting resources on inactive connections. And yes, we can get rid of this ping message, just need to think if it can be easily attacked somehow.